relayer/network_utils.go

```diff
@@ -0,0 +1,139 @@
+		if sourceBlockchain.GetSubnetID() == constants.PrimaryNetworkID {
+			if err := connectToPrimaryNetworkPeers(logger, network, cfg, sourceBlockchain); err != nil {
+				return fmt.Errorf(
+					"failed to connect to primary network peers: %w",
+					err,
+				)
+			}
+		} else {
+			if err := connectToNonPrimaryNetworkPeers(logger, network, cfg, sourceBlockchain); err != nil {
+				return fmt.Errorf(
+					"failed to connect to non-primary network peers: %w",
+					err,
+				)
+			}
+		}
+	}
+	return nil
+}
+
+// Connect to the validators of the source blockchain. For each destination blockchain,
+// verify that we have connected to a threshold of stake.
+func connectToNonPrimaryNetworkPeers(
+	logger logging.Logger,
+	network *peers.AppRequestNetwork,
+	cfg *config.Config,
+	sourceBlockchain *config.SourceBlockchain,
+) error {
+	subnetID := sourceBlockchain.GetSubnetID()
+	connectedValidators, err := network.ConnectToCanonicalValidators(subnetID)
+	if err != nil {
+		logger.Error(
+			"Failed to connect to canonical validators",
+			zap.String("subnetID", subnetID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+	for _, destination := range sourceBlockchain.SupportedDestinations {
+		blockchainID := destination.GetBlockchainID()
+		if ok, quorum, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); !ok {
+			logger.Error(
+				"Failed to connect to a threshold of stake",
+				zap.String("destinationBlockchainID", blockchainID.String()),
+				zap.Uint64("connectedWeight", connectedValidators.ConnectedWeight),
+				zap.Uint64("totalValidatorWeight", connectedValidators.TotalValidatorWeight),
+				zap.Any("warpQuorum", quorum),
+			)
+			return err
+		}
+	}
+	return nil
+}
+
+// Connect to the validators of the destination blockchains. Verify that we have connected
+// to a threshold of stake for each blockchain.
+func connectToPrimaryNetworkPeers(
+	logger logging.Logger,
+	network *peers.AppRequestNetwork,
+	cfg *config.Config,
+	sourceBlockchain *config.SourceBlockchain,
+) error {
+	for _, destination := range sourceBlockchain.SupportedDestinations {
+		blockchainID := destination.GetBlockchainID()
+		subnetID := cfg.GetSubnetID(blockchainID)
+		connectedValidators, err := network.ConnectToCanonicalValidators(subnetID)
+		if err != nil {
+			logger.Error(
+				"Failed to connect to canonical validators",
+				zap.String("subnetID", subnetID.String()),
+				zap.Error(err),
+			)
+			return err
+		}
+
+		if ok, quorum, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); !ok {
+			logger.Error(
+				"Failed to connect to a threshold of stake",
+				zap.String("destinationBlockchainID", blockchainID.String()),
+				zap.Uint64("connectedWeight", connectedValidators.ConnectedWeight),
+				zap.Uint64("totalValidatorWeight", connectedValidators.TotalValidatorWeight),
+				zap.Any("warpQuorum", quorum),
+			)
+			return err
+		}
+	}
+	return nil
```

Comment:
In the case that we don't connect to a threshold of stake, we log an error here and then return `nil`, is that's what is intended?

I think the intended flow here should be to check for an error first, return the error if present, then check `ok`, which should maybe return a new error if we haven't connected to enough stake.

Reply:
Thank you for finding this. It indeed seems like a bug in the existing implementation, but this whole initialization step is not really necessary I only kept it for now to maintain existing behavior. 

For now I updated the comment to indicate actual behavior, since I do think that the existing bug i.e. more permissive behavior is actually what we want but if we align on another approach I'm happy to implement it.